src/main/java/org/jenkinsci/plugins/workflow/graph/StandardGraphLookupView.java

```diff
@@ -0,0 +1,219 @@
+            if (overallEnclosing != null) {
+                nearestEnclosingBlock.put(newHead.getId(), overallEnclosing);
+            }
+        } else {
+            if (newHead instanceof BlockStartNode) {
+                blockStartToEnd.put(newHead.getId(), INCOMPLETE);
+            }
+
+            // Now we try to generate enclosing block info for caching, by looking at parents
+            // But we don't try to hard -- usually the cache is populated and we defer recursive walks of the graph
+            List<FlowNode> parents = newHead.getParents();
+            if (parents.size() > 0) {
+                FlowNode parent = parents.get(0);  // Multiple parents only for end of a parallel, and then both are BlockEndNodes
+
+                if (parent instanceof BlockStartNode) {
+                    nearestEnclosingBlock.put(newHead.getId(), parent.getId());
+                } else {
+                    // Try to reuse info from cache for this node:
+                    //   If the parent ended a block, we use info from the start of the block since that is at the same block nesting level as our new head
+                    //   Otherwise the parent is a normal atom node, and the head is enclosed by the same block
+                    String lookupId = (parent instanceof BlockEndNode) ? ((BlockEndNode) parent).getStartNode().getId() : parent.getId();
+                    String enclosingId = nearestEnclosingBlock.get(lookupId);
+                    if (enclosingId != null) {
+                        nearestEnclosingBlock.put(newHead.getId(), enclosingId);
+                    }
+                }
+            }
+        }
+    }
+
+    /** Create a lookup view for an execution */
+    public StandardGraphLookupView() {
+    }
+
+    @Override
+    public boolean isActive(@Nonnull FlowNode node) {
+        if (node instanceof FlowEndNode) { // cf. JENKINS-26139
+            return !node.getExecution().isComplete();
+        } else if (node instanceof BlockStartNode){  // BlockStartNode
+            return this.getEndNode((BlockStartNode)node) == null;
+        } else {
+            return node.getExecution().isCurrentHead(node);
+        }
+    }
+
+    // Do a brute-force scan for the block end matching the start, caching info along the way for future use
+    BlockEndNode bruteForceScanForEnd(@Nonnull BlockStartNode start) {
+        DepthFirstScanner scan = new DepthFirstScanner();
+        scan.setup(start.getExecution().getCurrentHeads());
+        for (FlowNode f : scan) {
+            if (f instanceof BlockEndNode) {
+                BlockEndNode end = (BlockEndNode)f;
+                BlockStartNode maybeStart = end.getStartNode();
+                // Cache start in case we need to scan again in the future
+                blockStartToEnd.put(maybeStart.getId(), end.getId());
+                if (start.equals(maybeStart)) {
+                    return end;
```

Comment:
Needs to be covered by a testcase (negative for coverage).

Reply:
Specifically: 
```
if (start.equals(maybeStart)) {
                    return end;
}
```

Other than that, line coverage seems okay (not 100% for StandardGraphLookupView but the cases uncovered are things like rethrowing exceptions and).